User controller testing

crod951 opened the pull request.

crod951 committed:
feat(user-controller-test): added user_controller_test for testing API routes from user_controller

crod951 committed:
feat(user-controller-test): added test for banned_address

crod951 committed:
refactor(user-controller-testing): added testing for already authenticated users

crod951 committed:
refactor(user-controller-test): added extra checking in asserts for errors; refactored asserts checking for user.id

unenglishable commented on test/epochtalk_server_web/controllers/user_controller_test.exs:
```diff
@@ -0,0 +1,232 @@
+  @invalid_login_username_attrs %{username: "test", password: "password"}
+
+
+  describe "check username and email" do
```
have route names/paths or method names in describe

what parts of the project are we testing?
perhaps split out tests that don't fall under the same method or route

unenglishable replied:
checking out [this blog](https://brooklinmyers.medium.com/phoenix-elixir-testing-beyond-mix-test-5b07de241001)...  i think the descriptions in this example are pretty useful.

![example](https://miro.medium.com/max/4800/1*5hA3TJ5s2rQSiWCUXQ6-xg.webp)

here, they note the tested functions and desired outcome.  this one shows the function with arity `(/2)`, which specifies exactly which version of the function they're testing.

unenglishable commented on test/epochtalk_server_web/controllers/user_controller_test.exs:
```diff
@@ -0,0 +1,232 @@
+  describe "check username and email" do
+    setup [:create_user]
+
+    test "check if username is taken", %{conn: conn, user: user} do
```
in test name, describe the action and desired outcome

what are we checking for, and what should the code do given particular input?

unenglishable commented on test/epochtalk_server_web/controllers/user_controller_test.exs:
```diff
@@ -0,0 +1,232 @@
+      conn = get(conn, Routes.user_path(conn, :email, user.email))
+      assert %{"found" => true} = json_response(conn, 200)
+    end
+  end
```
a solid test suite will check all code paths to ensure that the code works as expected in each case

consider the different possible starting conditions and what the results should be

also consider if there are any edge cases and test those as well

these suggestions apply to all tests

unenglishable replied:
for example:
in this test, we're checking email and username routes.  your test checks the functionality if username and email are already taken; it should also check if the functions return the right result when they aren't taken as well.

unenglishable committed:
Merge remote-tracking branch 'origin/main' into user-controller-testing-changes

unenglishable commented on test/epochtalk_server_web/controllers/user_controller_test.exs:
```diff
@@ -0,0 +1,232 @@
+      assert user.id == String.to_integer(json_response(conn, 200)["id"])
+    end
+
+    test "renders error when user is not loegged in", %{conn: conn} do
```
* logged in

unenglishable committed:
test(doctest): change User, Role -> Mention, Invitation

this is not really a fix, just switches to doctesting mention,
invitation, which are not seeded

need to actually test these later

unenglishable committed:
test(user-controller-authenticate): remove string to int conversion

on json_response, user id should be an int already

unenglishable committed:
Merge pull request #14 from epochtalk/user-controller-testing-changes

User controller testing changes

unenglishable commented on test/epochtalk_server_web/controllers/user_controller_test.exs:
```diff
@@ -0,0 +1,232 @@
+    end
+  end
+
+  describe "confirm" do
```
nice work here, we can tell you read the code and know how it works 👍

unenglishable reviewed (changes requested):
looks good so far, just some minor improvements to be made;

extra cases for checking fail/success to check more code paths

better naming will help us understand test failures at a glance, letting us know what's being tested and what the expected result is.  some of this is revealed by how elixir handles returning test results (says which line errored and what the expected result was), but it's also nice to have the descriptions match up

after these quick fixes, we'll merge in and talk about the next steps in a meet.  nice work!

unenglishable committed:
refactor: mix format

unenglishable committed:
chore(github-actions): publish redis port

redis should now be reachable on localhost for testing

unenglishable committed:
test(user controller): modify banned address seed

use only ip, not also hostname

banned address table only stores hostname or ip;
hostname takes precedence...  on CI, hostname check was failing

TODO: seed banned addresses properly

unenglishable committed:
test(user controller): debug malicious ip handling hostname

unenglishable committed:
test(bannedaddr): seed banned ip+addr, check for both in test

unenglishable committed:
style(user controller test): remove whitespace

unenglishable committed:
test(debug): remove debug prints

unenglishable committed:
style(mix format): run mix format

unenglishable committed:
test(debug): logs

unenglishable committed:
fix(bannedaddress): use inet instead of inet_res

https://elixirforum.com/t/inet-res-gethostbyaddr-locally/17092/2

localhost doesnt resolve in github actions

unenglishable committed:
revert: "test(debug): logs"

This reverts commit 72998db10f66f2791a83cf6cb4c3463e5ca72aaa.

unenglishable committed:
Merge pull request #15 from epochtalk/github-actions

GitHub actions
